### PR TITLE
throw RuntimeErrors for mkldnn modules when mkldnn is disabled

### LINKED ISSUE #40391: Mkldnn modules should throw RuntimeErrors if invoked when mkldnn is disabled
Currently the following script will succeed:
```
import torch
import copy
from torch.utils import mkldnn as mkldnn_utils
groups=1
N = 32
C = 32
M = 32
x = torch.randn(N, C, 56, 56, dtype=torch.float).to_mkldnn()
conv2d = torch.nn.Conv2d(in_channels=C,
                            out_channels=M,
                            kernel_size=3,
                            padding=1)
mkldnn_conv2d = mkldnn_utils.to_mkldnn(copy.deepcopy(conv2d))
with torch.backends.mkldnn.flags(enabled=False):
    y_mkldnn = mkldnn_conv2d(x).to_dense()
```
It is confusing that when mkldnn is disabled, one can still call mkldnn functions. An attempt to do so should throw a RuntimeError.

### PR DESCRIPTION
**Pull request by XiaobingSuper**
fix https://github.com/pytorch/pytorch/issues/40391, we will throw error for mkldnn module when mkldnn disabled.

### CODE REVIEW

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -618,6 +618,9 @@ at::Tensor _convolution(
     bool benchmark, bool deterministic, bool cudnn_enabled) {
 
   const bool input_is_mkldnn = input_r.is_mkldnn();
+  if (input_is_mkldnn) {
```
Comment (ngimel) [inline aten/src/ATen/native/Convolution.cpp:621]: Is it enough to check only for mkldnn inputs? Currently the following (which, as far as I understand, sends regular tensor to mkldnn convolution) works:
```
x = torch.randn(N, C, 56, 56, dtype=torch.float)#.to_mkldnn()
conv2d = torch.nn.Conv2d(in_channels=C,
                            out_channels=M,
                            kernel_size=3,
                            padding=1)
mkldnn_conv2d = mkldnn_utils.to_mkldnn(copy.deepcopy(conv2d))
with torch.backends.mkldnn.flags(enabled=True):
    y_mkldnn = mkldnn_conv2d(x)#.to_dense()
```
Same question for all other functions - you added checks only for mkldnn inputs.

Comment (XiaobingSuper) [reply aten/src/ATen/native/Convolution.cpp:621]: @ngimel , for convolution, this input check can be removed, see
1. torch.backends.mkldnn.flags(enabled=True), input and weight are MKLNDN tensor, it is ok.
2. torch.backends.mkldnn.flags(enabled=True), input is dense tensor, and weight are MKLNDN tensor, will call MKLDNN path, there will make check to make sure input and weight have same data type: see 
https://github.com/pytorch/pytorch/blob/6468bc46375e4cb73fe3b59c7b81dff559bb0967/aten/src/ATen/native/Convolution.cpp#L746-L748, so this will report a error: **Input type (torch.FloatTensor) and weight type (Mkldnntorch.FloatTensor) should be the same**
3. torch.backends.mkldnn.flags(enabled=False), input and weight are MKLNDN tensor,  will call this path 
https://github.com/pytorch/pytorch/blob/6468bc46375e4cb73fe3b59c7b81dff559bb0967/aten/src/ATen/native/Convolution.cpp#L801-L802, which not support MKLDNN inputs. So this will report a error: **"opaque tensors do not have is_contiguous"**.
4.  torch.backends.mkldnn.flags(enabled=False), input is dense tensor,  weight are MKLNDN tensor, also call path 
https://github.com/pytorch/pytorch/blob/6468bc46375e4cb73fe3b59c7b81dff559bb0967/aten/src/ATen/native/Convolution.cpp#L801-L802
this will report a error: **Cannot access data pointer of Tensor that doesn't have storage**

File: aten/src/ATen/native/Linear.cpp [hunk 2]
```diff
@@ -15,6 +15,14 @@ namespace at { namespace native {
...
   if (input.is_mkldnn()) {
+    TORCH_CHECK(at::globalContext().userEnabledMkldnn(),
+        "MKLDNN linear module can't run when mkldnn disabled");
+    TORCH_CHECK(input.options().type_equal(weight.options()),
+        "Input type (", input.toString(), ") and weight type (", weight.toString(),
+        ") should be the same");
+    TORCH_CHECK(!bias.defined() || (input.options().type_equal(bias.options())),
+        "Input type (", input.toString(), ") and bias type (", bias.toString(),
+        ") should be the same");
     return at::mkldnn_linear(input, weight, bias);
```
Comment (XiaobingSuper) [inline aten/src/ATen/native/Linear.cpp:26]: For linear:
```
x = torch.randn(56, 56, dtype=torch.float)
linear = torch.nn.Linear(56, 56 bias=True)
mkldnn_linear = mkldnn_utils.to_mkldnn(copy.deepcopy(linear))
with torch.backends.mkldnn.flags(enabled=True):
    y_mkldnn = mkldnn_linear(x)
```
1. torch.backends.mkldnn.flags(enabled=True), input and weight are MKLDNN tensor, it can works.
2. torch.backends.mkldnn.flags(enabled=True), input is dense tensor,  weight are MKLDNN tensor, it can works, because we will first transfer input to MLDNN tensor in https://github.com/pytorch/pytorch/blob/bdc00196d1b63a339a5b9b6aaec44cefe069a6ea/torch/utils/mkldnn.py#L30.
3.  torch.backends.mkldnn.flags(enabled=False), input and weight are MKLDNN tensor, it can report error: **MKLDNN linear module can't run when mkldnn disabled**
4. torch.backends.mkldnn.flags(enabled=False), input is dense tensor, weight are MKLDNN tensor, it can also report error: **MKLDNN linear module can't run when mkldnn disabled**.

File: aten/src/ATen/native/Normalization.cpp [hunk 3]
```diff
@@ -463,6 +463,11 @@ std::tuple<Tensor, Tensor, Tensor, Tensor, int64_t> _batch_norm_impl_index(
              std::make_tuple(2));
   }
 
+  if (input.is_mkldnn()) {
+    TORCH_CHECK(at::globalContext().userEnabledMkldnn(),
+        "MKLDNN batch_norm module can't run when mkldnn is disabled");
+  }
+
```
Comment (XiaobingSuper) [inline aten/src/ATen/native/Normalization.cpp:470]: For BatchNorm,
```
x = torch.randn(1, 3, 56, 56, dtype=torch.float)
bn = torch.nn.BatchNorm2d(3).float().train(False)
mkldnn_bn = mkldnn_utils.to_mkldnn(copy.deepcopy(bn))
with torch.backends.mkldnn.flags(enabled=True):
    y_mkldnn = bn(x)
```
1. torch.backends.mkldnn.flags(enabled=True), input and weight are MKLDNN tensor, it can works.
1. torch.backends.mkldnn.flags(enabled=True), input is dense tensor,  weight are MKLDNN tensor, it report a error:**mkldnn_tensor.is_mkldnn() INTERNAL ASSERT FAILED at "../aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp":56, please report a bug to PyTorch. mkldnn_to_dense expects MKL-DNN tensor input**.
3. torch.backends.mkldnn.flags(enabled=False), input and weight are MKLDNN tensor, it will report a error:**MKLDNN batch_norm module can't run when mkldnn is disabled**.
4. torch.backends.mkldnn.flags(enabled=False), input is dense tensor, weight are MKLDNN tensor, it will report a error:**mkldnn_tensor.is_mkldnn() INTERNAL ASSERT FAILED at "../aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp":56, please report a bug to PyTorch. mkldnn_to_dense expects MKL-DNN tensor input**.

File: torch/utils/mkldnn.py [hunk 4]
```diff
@@ -59,13 +59,15 @@ def __getstate__(self):
 
     @torch.jit.script_method
     def forward(self, x):
-        return torch.mkldnn_convolution(
+        return torch.convolution(
```
[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/utils/mkldnn.py:62]: enable it at https://github.com/pytorch/pytorch/pull/40483

File: aten/src/ATen/native/Normalization.cpp (hunk 3 above)
Comment (ngimel) [reply aten/src/ATen/native/Normalization.cpp:470]: There should not be INTERNAL_ASSERTs here, right? Is it a real pytorch bug that you guys have to fix, or should the error message be different. Couple additional questions
1) why is mkldnn_to_dense attempted here, is it when converting output of the operation to dense tensor?
2) Why is behavior here different than in linear, where case 2 works, and case 4 errors out as expected, with expected error message?

File: torch/utils/mkldnn.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/utils/mkldnn.py:62]: what is convolution behavior wrt mkl or dense inputs? Similar to cases you posted above for linear and bn.

File: aten/src/ATen/native/Normalization.cpp (hunk 3 above)
Comment (XiaobingSuper) [reply aten/src/ATen/native/Normalization.cpp:470]: For the first one, it call in https://github.com/pytorch/pytorch/blob/7b0f867c4853af69b14bb516ba2a9dbff8f56d51/aten/src/ATen/native/mkldnn/MKLDNNCommon.cpp#L54, which the error message is wrong, it shold be "itensor_from_mkldnn expects MKL-DNN tensor input"

For the second: the case 2 can be works for linear is that we always convert dense tensor to MKLDNN tensor for input, see https://github.com/pytorch/pytorch/blob/7b0f867c4853af69b14bb516ba2a9dbff8f56d51/torch/utils/mkldnn.py#L30. the case 4's error message need make some changes to be unified.

File: torch/utils/mkldnn.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ngimel) [reply torch/utils/mkldnn.py:62]: After #40610 torch.convolution won't be calling mkldnn in a fair number of cases, so you still need to call mkldnn_convolution here.

[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/utils/mkldnn.py:62]: @ngimel, I don't think need to call mkldnn_convolution here, if input and weight are MKLDNN tensor, it will call mkldnn_convolution directly, see https://github.com/pytorch/pytorch/blob/fcadca1bdab62ea5728e49397e023c8863466e83/aten/src/ATen/native/Convolution.cpp#L230-L239. if users set input to MKLDNN tensor, assuming they want to call MKLDNN path.

[the remark this answers is not in the document]
Comment (XiaobingSuper) [reply torch/utils/mkldnn.py:62]: @ngimel
